### PR TITLE
Add cloud exercise

### PR DESCRIPTION
**Pull request by haroal**
I wrote a basic tech exercise to test candidates applying for a job in the Cloud team.

It is quite simple, but will demonstrate the ability to:
- use Git
- use Docker/docker-compose
- use Django
- use libraries (`django-rest-framework`)
- write good commented, well-organized Python code
- create API (with correct HTTP status codes)
- create a model
- use serializers
- write an algorithm

And should take less than 1 hour (I made it in 30 minutes but I had the setup already made).

### CODE REVIEW

File: cloud/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+*It should take less than 1 hour to complete. If you think some parts need more, please tell us.*
+
+## Description
+
+The goal of this exercise is to build an API that will allow a user to register time blocks with a name
+and then query the resulting timeline.
+
+1. Setup a Django server within a Docker container. This server will need a PostgreSQL database to store models, which
+   will also be into a Docker container. We recommend describing these containers using docker-compose.
```
Comment (mattrunchey) [inline cloud/README.md:23]: I think it is helpful to provide more of a framework/skeleton of the project here. Unless you require them to intimately remember off-hand how to make new Dockerfiles and compose files, having the "basic" one present may shave off time and keep candidates moving forward.

So, for that I would suggest providing a Dockerfile and `docker-compose.yml` and `.env`, with the bare essentials, perhaps with comments like `#TODO` indicating where we expect candidates to complete/include their logical solution. 
- I would say at a bare minimum, providing the Dockerfile will be helpful - at the very least it ensures that you are using the same Django version/image as they are!

File: cloud/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+2. Define the `Timeblock` model and migrations to save it in DB.
```
Comment (mattrunchey) [inline cloud/README.md:24]: Similarly, do you think it makes sense to provide the template/boilerplate Django project folder (like `server/*`)? Perhaps even include the `Timeblock` model stub where you expect it to be. 

The motivation behind these comments is to give you a more "standard" solution to look at. One thing that I often find helps candidates as well is to provide them a basic set of unit tests that they can run to make sure their solution covers "the basics". I have my own private larger test suite that exercises corner cases and catches errors in algorithms, but that depends on a consistent structure and input/output to the system.

File: cloud/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   - id: int (auto-increment)
+   - name: str
+   - start_time: datetime
+   - end_time: datetime
+3. Add API routes using the `django-rest-framework` library. You have to choose the names and methods.
```
Comment (mattrunchey) [inline cloud/README.md:29]: And here, I think it doesn't tell the candidate too much to show a skeleton of where you want them to put their routes. They will still need to modify all of the right files (views, urls, etc).

**Review by mattrunchey — COMMENTED** [review]
LGTM

File: cloud/README.md (hunk 1 above)
Comment (mattrunchey) [reply cloud/README.md:23]: Example structure might be the way I built up the `radiation` folder. I remember David thinking that I made it too easy for candidates, but plenty of them still had trouble with docker, and the ones that didn't were able to give me interesting solutions that were cool even with all that I provided them :)

**Comment by haroal** [comment]
@mattrunchey 

> candidates may not have their own personal laptop or desktop environments set up to use Docker

I can ask during the intro call to check if they have access to the needed material. If not, I could make a tech call instead.

> a challenge like this pretty much will rule out anyone who doesn't know how to create docker containers very well

In 2022, a "Senior Backend Software Engineer" who doesn't know how to use Docker is definitely not as strong as what we are looking for. It would be a big red flag for me. Using Docker is part of the job and even if the person hasn't ever used it professionnaly (big doubt about this), not knowing how to use it would be a lack of personnal curiosity.

> Unless you require them to intimately remember off-hand how to make new Dockerfiles and compose files

There are a lot of resources on Google to create a Dockerfile/docker-compose file without knowing it by heart. And being able to find the right information on Google is also a good skill. 

> The motivation behind these comments is to give you a more "standard" solution to look at.

Providing a complete skeleton would make all the submitted exercises look the same and so we will lose a lot of information about the candidate: I think that not having a "standard" solution is a strength. Letting him/her build the project from scratch will show us how organized he/she is. Then during the part 2 call I could ask why the candidate choose to do it this way and not the way I was expecting him/her to do.

> One thing that I often find helps candidates as well is to provide them a basic set of unit tests

I agree about providing some tests to validate their work. I changed my mind because I wanted them to write the tests.

However, I think it is a good idea to write my own test suite to run it on their code and check for some edge cases they might have forgotten (even while writting their own tests) :)

Validating their work would be:
- looking into the overall structure
- running their tests
- running my tests
- looking at the 4 key points of the exercise: docker files, API routes, model and timeline algorithm

**Conclusion**

I think this exercise is easy enough not to provide any skeleton. I'm just asking for a very basic Django API and you have plenty of resources to build one on Google...
Being able to build something from scratch is part of the basic requirements we expect for a senior dev. If, afterwards, candidates complain about how hard it was, I will add more files to help them but I don't want them just to fill in the blancks. I want to know how the candidate would organize its project (files hierarchy, ...etc) because it is a good overview of the way the candidate is actually thinking. Then, Django and `django-rest-framework` already provides skeletons and files examples (it already is a framework after all ^^). 

So I would be more in a logic of starting with minimal information and add some according to candidates feedbacks. If you worry about the time needed to complete the exercise, I can discuss with the candidate at the end of the intro call to know if he/she agrees to make it or if he/she wants me to reduce its complexity (by adding some files for instance).

**Comment by haroal** [comment]
This exercise looks like a Jira ticket describing a feature the candidate would have to handle if he/she is hire :) (actually it is more detailed than most of our Jira tickets, but communications is not as easy so... :smile: )

**Review by mattrunchey — APPROVED** [review]
OK, sounds like you have a solid idea of the signal you're seeking. Let's see how it goes :)

**Comment by mattrunchey** [comment]
(one note -- we might want to "Delete" this PR convo, as candidates could peruse around and look at this)

**Comment by haroal** [comment]
Sure, let's delete our messages here :)
